actix/src/sec-7-stream.md

```diff
@@ -1 +1,305 @@
+        // Create server listener
+        let addr = net::SocketAddr::from_str("127.0.0.1:4321").unwrap();
+        let listener = TcpListener::bind(&addr).unwrap();
+
+        AcceptServer::create(|ctx| {
+            // Add message stream allows taking events from a stream, mapping them
+            // into Messages, and then sending them to Actors for handling.
+            ctx.add_message_stream(listener.incoming().map_err(|_| ()).map(|st| {
+                let addr = st.peer_addr().unwrap();
+                // Everytime a new connection is made, the st (new accepted connection) is
+                // sent as a TcpConnect message to the AcceptServer Actor.
+                TcpConnect(st, addr)
+            }));
+            AcceptServer {}
+        });
+
+        println!("Running tcp echo server on 127.0.0.1:4321");
+    });
+}
+```
+
+## Using a stream as a source of messages
+
+Now that we are able to accept new connections, we need to do something with them.
+We can use our accepted connection as a stream of frames, where those frames are
+sent as messages to an actor. Each connection is associated to a unique actor.
+
+# Writing a Codec
+
+First, we need to write a codec capable of converting bytes into whole frames
+that will eventually become messages. Codecs have two haves - the Decoder, that takes
+user input and returns frames, and an encoder that takes frames and converts them
+to bytes to be transmitted to the client.
+
+An important aspect of our codec and it's ability to be used with Actix is that
+the frames must implement the trait Message. If the frame doesn't implement Message you may need
+to use `add_message_stream` instead to perform the frame to message encapsulation.
+
+
+```rust
+extern crate bytes;
+use bytes::{BufMut, BytesMut};
+
+struct Bytes(Vec<u8>);
+
+impl Message for Bytes {
+    type Result = ();
+}
+
+struct ByteCodec;
+
+impl Decoder for ByteCodec {
+    type Item = Bytes;
+    type Error = io::Error;
+
+    fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<Bytes>, io::Error> {
+        if buf.len() > 0 {
+            let len = buf.len();
+            let b = Bytes(buf.split_to(len).to_vec());
+            // We have a full frame, return it.
+            Ok(Some(b))
+        } else {
+            // We have no data, so return that we don't have enough to complete
+            // a full frame. It's important you handle this case.
+            Ok(None)
+        }
+    }
+}
+
+impl Encoder for ByteCodec {
+    type Item = Bytes;
+    type Error = io::Error;
+
+    fn encode(&mut self, data: Bytes, buf: &mut BytesMut) -> Result<(), io::Error> {
+        buf.reserve(data.0.len());
+        buf.put(data.0);
+        // Indicate that we correctly encoded the frame and it can now be transmitted.
+        Ok(())
+    }
+}
+```
+
+For more about the implementation of Codecs, see tokio's documentation about Frames https://tokio.rs/docs/going-deeper/frames/
+
+^ How to make this a proper link?
```

Comment:
`[name](link)`

Reply:
Thanks, will fix this and the other comments

Reply:
Don't forget remove this line